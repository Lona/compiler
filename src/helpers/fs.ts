import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
