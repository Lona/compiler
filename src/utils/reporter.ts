export type Reporter = {
  info(...args: any[]): void
  log(...args: any[]): void
  warn(...args: any[]): void
  error(...args: any[]): void
}

export const defaultReporter: Reporter = {
  info: console.info.bind(console),
  log: console.log.bind(console),
  warn: console.warn.bind(console),
  error: console.error.bind(console),
}

export const silentReporter: Reporter = {
  log: () => {},
  info: () => {},
  warn: () => {},
  error: () => {},
}
