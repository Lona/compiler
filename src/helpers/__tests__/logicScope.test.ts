import * as fs from 'fs'
import * as path from 'path'
import * as serialization from '@lona/serialization'
import * as LogicAST from '../logicAst'
import { build } from '../logicScope'
import { defaultReporter } from '../reporter'

let id = 0
jest.mock('uuid', () => ({ v4: () => '' + id++ }))

function buildScope(filePaths: string[]) {
  const programs = filePaths.map(
    x =>
      LogicAST.makeProgram(
        JSON.parse(
          serialization.convertLogic(
            fs.readFileSync(path.join(__dirname, x), 'utf8'),
            serialization.SERIALIZATION_FORMAT.JSON
          )
        )
      ) as LogicAST.AST.Program
  )

  return {
    programs,
    scope: build(
      filePaths.map((x, i) => ({ node: programs[i], in: x })),
      defaultReporter
    ),
  }
}

describe('Scope', () => {
  beforeEach(() => {
    id = 0
  })

  test('simple reference', () => {
    const { scope, programs } = buildScope([
      './fixtures/simple-reference.logic',
    ])

    expect(Object.keys(scope.identifierToPattern).length).toBe(1)

    const aIdentifier = Object.keys(scope.identifierToPattern)[0]

    expect(
      (LogicAST.findNode(programs[0], aIdentifier) as LogicAST.AST.Identifier)
        .string
    ).toBe('a')
    expect(
      (LogicAST.findNode(
        programs[0],
        scope.identifierToPattern[aIdentifier].pattern
      ) as LogicAST.AST.Pattern).name
    ).toBe('a')

    expect(
      scope.patternToIdentifier[scope.identifierToPattern[aIdentifier].pattern]
        .identifier
    ).toBe(aIdentifier)

    expect(scope.identifierToPattern).toMatchSnapshot()
  })
})
