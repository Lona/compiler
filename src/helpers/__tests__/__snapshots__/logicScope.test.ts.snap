// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scope simple reference 1`] = `
Object {
  "21": Object {
    "in": "./fixtures/simple-reference.logic",
    "pattern": "0",
  },
}
`;
